### PR TITLE
[TEST] Don't allow constant eval to access generator layouts

### PR DESCRIPTION
**Pull request by Zoxc**
This emits an error if generator layouts are observed by the frontend via constant evaluation. This ensures that the frontend does not depend on optimizations done later in MIR. This is a breaking change as generator layouts can be observed via the async feature combined with constant evaluation.

r? @oli-obk

### CODE REVIEW

File: compiler/rustc_const_eval/src/const_eval/machine.rs [hunk 1]
```diff
@@ -357,6 +357,9 @@ impl<'mir, 'tcx> interpret::Machine<'mir, 'tcx> for CompileTimeInterpreter<'mir,
 
     const PANIC_ON_ALLOC_FAIL: bool = false; // will be raised as a proper error
 
+    // We don't allow access to generator layout at compile time.
+    const ACCESS_GENERATOR_LAYOUT: bool = false;
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/const_eval/machine.rs:361]: Please remove the default value here, and move it to the compile_time_machine macro below. After all this is also shared with Miri, so we don't set defaults here based on "being compile-time".

File: compiler/rustc_const_eval/src/interpret/validity.rs [hunk 2]
```diff
@@ -950,4 +950,16 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         // recurse through references which, for now, we don't want here, either.
         self.validate_operand_internal(op, vec![], None, None)
     }
+
+    /// This function checks if the given `ty`'s layout depends on generators.
+    #[inline(always)]
+    pub fn validate_generator_layout_access(&self, ty: Ty<'tcx>) -> InterpResult<'tcx> {
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/validity.rs:956]: "validity.rs" is for checking that a value is valid. Please don't put unrelated functions here.

File: compiler/rustc_middle/src/mir/interpret/error.rs [hunk 3]
```diff
@@ -404,6 +404,7 @@ pub enum ValidationErrorKind<'tcx> {
     InvalidBool { value: String },
     InvalidChar { value: String },
     InvalidFnPtr { value: String },
+    GeneratorLayoutAccess { ty: Ty<'tcx>, generators: &'tcx List<Ty<'tcx>> },
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/interpret/error.rs:407]: "ValidationErrorKind" is for when a value doesn't satisfy its validity invariant. This is unrelated, please don't put that here.

Sound more like an `InvalidProgram` error to me. Or ideally the error variant is only added in const_eval::error, so we structurally know it will not occur in Miri.

File: src/tools/miri/src/machine.rs [hunk 4]
```diff
@@ -1410,17 +1411,14 @@ impl<'mir, 'tcx> Machine<'mir, 'tcx> for MiriMachine<'mir, 'tcx> {
...
         let Some(Provenance::Concrete { alloc_id, .. }) = mplace.ptr.provenance else {
             panic!("after_local_allocated should only be called on fresh allocations");
         };
         let local_decl = &ecx.active_thread_stack()[frame].body.local_decls[local];
         let span = local_decl.source_info.span;
-        ecx.machine
-            .allocation_spans
-            .borrow_mut()
-            .insert(alloc_id, (span, None));
+        ecx.machine.allocation_spans.borrow_mut().insert(alloc_id, (span, None));
```
Comment (RalfJung) [inline src/tools/miri/src/machine.rs:1421]: Please don't format these files, they'll have to be reformatted when syncing changes back to Miri. (Miri has a different rustfmt.toml.)

**Comment by oli-obk** [comment]
> This is a breaking change as generator layouts can be observed via the async feature combined with constant evaluation.

do you have an example snippet that compiles on stable but doesn't compile anymore?

**Comment by Zoxc** [comment]
As @RalfJung pointed out, this isn't the only place layout is used. It would have apply more generally to all allocations, effectively banning async and generator blocks from const eval, at least while allocations are byte based. That's not a particularly great outcome.

@oli-obk I'll relay your example back to you:
```rust
const X: usize = {
    let x = Thing(|x: u128| async move {x});
    x.size()
};

struct Thing<T>(T);

impl<U, T: Fn(u128) -> U> Thing<T> {
    const fn size(&self) -> usize {
        std::mem::size_of::<U>()
    }
}
```

**Comment by oli-obk** [comment]
Oh whoops. I remember this.

Yea, this seems very annoying to prevent. I forget where the original discussion was. Why did we consider avoiding getting generator layouts during const eval? Avoiding some cycle errors?

**Comment by Zoxc** [comment]
[Here's the discussion on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Inter-procedural.20opts.20before.20generator.20lowering).

The problem is that having const eval depend on generator layouts places constraints on optimization. We can't have an out of process LTO optimization of generator layouts for example. Just partially optimizing generator layouts within the query system gets quite hairy. I suspect there's some interesting interactions with monomorphization too.

**Comment by oli-obk** [comment]
Thanks! I reread the thread. Splitting Reveal::All into `Optimizations` and `Codegen` indeed seems like the only solution to do this properly. We can probably get away with letting const eval always use `Codegen` for its Allocations, but forward its real `Reveal` otherwise. `layout_of` with `Reveal::Optimizations` or `UserFacing` would then error in case it tries to compute the layout of a generator

**Comment by RalfJung** [comment]
Inevitably at some point people will ask to execute generators at compiletime. What do we do then?

I don't think doing LTO on pre-lowering generators is a realistic option.

File: src/tools/miri/src/machine.rs (hunk 4 above)
Comment (Zoxc) [reply src/tools/miri/src/machine.rs:1421]: Why doesn't rustfmt use those settings? :/

Comment (RalfJung) [reply src/tools/miri/src/machine.rs:1421]: It's up to RA to tell it the config file to use (rustfmt isn't invoked on a file, it is fed data via stdin and returned via stdout). Not sure if there's an open RA bug for this.

Comment (Zoxc) [reply src/tools/miri/src/machine.rs:1421]: It seems like [RA changes the current directory](https://github.com/rust-lang/rust-analyzer/pull/7371) to match the file for `rustfmt`. I guess that doesn't work out.

**Comment by oli-obk** [comment]
The moment `offset_of` works on stable in const eval, I don't see how we could possibly hide the actual size of generators. Without `offset_of`, we can always do some shenanigans where the size of a generator is treated as something opaque just like pointers. That would be a lot of complexity, but it would be possible. Since this will break down once we have `offset_of`, there's no real point to it though.

the new Reveal mode would resolve the issue raised in the zulip, as all it would do is make optimizations before the generator lowering pass stop being able to rely on the size of generators (without causing errors, just by returning something akin to `TooGeneric`, which all optimizations must already silently handle)

**Comment by RalfJung** [comment]
https://github.com/Gilnaa/memoffset already works in const-eval when there is no interior mutability.

**Comment by apiraino** [comment]
hey, checking progress. I'm unsure about the status of this PR (besides a few outstanding comments and a rebase). @Zoxc do you have any update? Are you specifically waiting on a feedback? Thanks!

@rustbot  author

**Comment by Zoxc** [comment]
@oli-obk The problem with that approach is that we end up having that reveal option on anything that could use generator sizes, including `type_of` and type checking.

I don't think the method in this PR is a suitable solution however, so I'm going to close it.
